File: clients/android/app/src/main/kotlin/com/lextures/android/core/lms/LmsFeatureModelsFeed.kt
package com.lextures.android.core.lms

import kotlinx.serialization.Serializable

// region Course feed & channels (M7.6)

@Serializable
data class FeedChannel(
    val id: String = "",
    val name: String = "",
    val sortOrder: Int = 0,
    val createdAt: String = "",
)

@Serializable
data class FeedMessage(
    val id: String = "",
    val channelId: String = "",
    val authorUserId: String = "",
    val authorEmail: String = "",
    val authorDisplayName: String? = null,
    val parentMessageId: String? = null,
    val body: String = "",
    val mentionsEveryone: Boolean = false,
    val mentionUserIds: List<String> = emptyList(),
    val pinnedAt: String? = null,
    val createdAt: String = "",
    val editedAt: String? = null,
    val likeCount: Int = 0,
    val viewerHasLiked: Boolean = false,
    val replies: List<FeedMessage> = emptyList(),
) {
    val authorLabel: String get() = authorDisplayName?.takeIf { it.isNotBlank() } ?: authorEmail
}

@Serializable
data class FeedRosterPerson(
    val userId: String = "",
    val email: String = "",
    val displayName: String? = null,
    val avatarUrl: String? = null,
) {
    val label: String get() = displayName?.takeIf { it.isNotBlank() } ?: email
}

@Serializable
data class FeedImageUpload(
    val id: String = "",
    val content_path: String = "",
    val mime_type: String = "",
    val byte_size: Long = 0,
)

@Serializable
data class FeedChannelsResponse(val channels: List<FeedChannel> = emptyList())

@Serializable
data class FeedMessagesResponse(val messages: List<FeedMessage> = emptyList())

@Serializable
data class FeedRosterResponse(val people: List<FeedRosterPerson> = emptyList())

@Serializable
data class CreateFeedChannelBody(val name: String)

@Serializable
data class PostFeedMessageBody(
    val body: String,
    val parentMessageId: String? = null,
    val mentionUserIds: List<String> = emptyList(),
    val mentionsEveryone: Boolean = false,
)

@Serializable
data class PatchFeedMessageBody(val body: String)

@Serializable
data class PinFeedMessageBody(val pinned: Boolean)

@Serializable
data class PostFeedMessageResponse(val id: String = "")

// endregion

File: clients/ios/Lextures/Core/LMS/LMSFeatureModelsFeed.swift
import Foundation

// MARK: - Course feed & channels (M7.6)

struct FeedChannel: Codable, Identifiable, Hashable {
    var id: String
    var name: String
    var sortOrder: Int
    var createdAt: String
}

struct FeedMessage: Codable, Identifiable, Hashable {
    var id: String
    var channelId: String
    var authorUserId: String
    var authorEmail: String
    var authorDisplayName: String?
    var parentMessageId: String?
    var body: String
    var mentionsEveryone: Bool
    var mentionUserIds: [String]
    var pinnedAt: String?
    var createdAt: String
    var editedAt: String?
    var likeCount: Int
    var viewerHasLiked: Bool
    var replies: [FeedMessage]

    var authorLabel: String {
        authorDisplayName?.isEmpty == false ? authorDisplayName! : authorEmail
    }

    init(
        id: String,
        channelId: String,
        authorUserId: String,
        authorEmail: String,
        authorDisplayName: String?,
        parentMessageId: String?,
        body: String,
        mentionsEveryone: Bool,
        mentionUserIds: [String],
        pinnedAt: String?,
        createdAt: String,
        editedAt: String?,
        likeCount: Int,
        viewerHasLiked: Bool,
        replies: [FeedMessage]
    ) {
        self.id = id
        self.channelId = channelId
        self.authorUserId = authorUserId
        self.authorEmail = authorEmail
        self.authorDisplayName = authorDisplayName
        self.parentMessageId = parentMessageId
        self.body = body
        self.mentionsEveryone = mentionsEveryone
        self.mentionUserIds = mentionUserIds
        self.pinnedAt = pinnedAt
        self.createdAt = createdAt
        self.editedAt = editedAt
        self.likeCount = likeCount
        self.viewerHasLiked = viewerHasLiked
        self.replies = replies
    }

    // The server can send `null` (not `[]`) for `mentionUserIds`/`replies` when a message has
    // neither — decode those leniently as empty arrays rather than failing the whole payload.
    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        id = try container.decode(String.self, forKey: .id)
        channelId = try container.decode(String.self, forKey: .channelId)
        authorUserId = try container.decode(String.self, forKey: .authorUserId)
        authorEmail = try container.decode(String.self, forKey: .authorEmail)
        authorDisplayName = try container.decodeIfPresent(String.self, forKey: .authorDisplayName)
        parentMessageId = try container.decodeIfPresent(String.self, forKey: .parentMessageId)
        body = try container.decode(String.self, forKey: .body)
        mentionsEveryone = try container.decode(Bool.self, forKey: .mentionsEveryone)
        mentionUserIds = try container.decodeIfPresent([String].self, forKey: .mentionUserIds) ?? []
        pinnedAt = try container.decodeIfPresent(String.self, forKey: .pinnedAt)
        createdAt = try container.decode(String.self, forKey: .createdAt)
        editedAt = try container.decodeIfPresent(String.self, forKey: .editedAt)
        likeCount = try container.decode(Int.self, forKey: .likeCount)
        viewerHasLiked = try container.decode(Bool.self, forKey: .viewerHasLiked)
        replies = try container.decodeIfPresent([FeedMessage].self, forKey: .replies) ?? []
    }
}

struct FeedRosterPerson: Codable, Identifiable, Hashable {
    var userId: String
    var email: String
    var displayName: String?
    var avatarUrl: String?

    var id: String { userId }

    var label: String {
        displayName?.isEmpty == false ? displayName! : email
    }
}

struct FeedImageUpload: Decodable {
    var id: String
    var contentPath: String
    var mimeType: String
    var byteSize: Int64

    enum CodingKeys: String, CodingKey {
        case id
        case contentPath = "content_path"
        case mimeType = "mime_type"
        case byteSize = "byte_size"
    }
}

struct FeedChannelsResponse: Decodable {
    var channels: [FeedChannel]
}

struct FeedMessagesResponse: Decodable {
    var messages: [FeedMessage]
}

struct FeedRosterResponse: Decodable {
    var people: [FeedRosterPerson]
}

struct CreateFeedChannelRequest: Encodable {
    var name: String
}

struct PostFeedMessageRequest: Encodable {
    var body: String
    var parentMessageId: String?
    var mentionUserIds: [String]
    var mentionsEveryone: Bool
}

struct PatchFeedMessageRequest: Encodable {
    var body: String
}

struct PinFeedMessageRequest: Encodable {
    var pinned: Bool
}

struct PostFeedMessageResponse: Decodable {
    var id: String
}
